From jasonuhl@sgi.com Mon Mar  7 19:26:19 2005
Date: Mon, 7 Mar 2005 19:26:19 -0800
From: Jason Uhlenkott <jasonuhl@sgi.com>
To: lkcd-devel@lists.sourceforge.net
Subject: [PATCH 6/7] Polling I/O: sata_vsc support
Message-ID: <20050308032619.GF103743@dragonfly.engr.sgi.com>
References: <20050308031638.GC98711@dragonfly.engr.sgi.com>
Mime-Version: 1.0
Content-Type: text/plain; charset=us-ascii
Content-Disposition: inline
In-Reply-To: <20050308031638.GC98711@dragonfly.engr.sgi.com>
User-Agent: Mutt/1.5.6i
Status: RO
Content-Length: 1308
Lines: 50

Polling I/O support for Vitesse VSC7174 4 port DPA SATA.

Index: linux/drivers/scsi/sata_vsc.c
===================================================================
--- linux.orig/drivers/scsi/sata_vsc.c	2005-03-07 17:48:57.889069897 -0800
+++ linux/drivers/scsi/sata_vsc.c	2005-03-07 17:49:09.440925190 -0800
@@ -186,6 +186,34 @@
 	return IRQ_RETVAL(handled);
 }
 
+static int
+vsc_dump_quiesce(struct scsi_device *device)
+{
+	struct ata_port *ap = (struct ata_port *) device->host->hostdata;
+	struct ata_host_set *host_set = ap->host_set;
+
+	/*
+	 * Give the device 1 second to finish whatever it might be doing,
+	 * then deal with any pending interrupts.
+	 */
+	mdelay(1000);
+	vsc_sata_interrupt(0, host_set, NULL);
+
+	ap->qactive = 0;
+	ap->active_tag = ATA_TAG_POISON;
+
+	return 0;
+}
+
+static void
+vsc_dump_poll(struct scsi_device *device)
+{
+	struct ata_port *ap = (struct ata_port *) device->host->hostdata;
+	struct ata_host_set *host_set = ap->host_set;
+
+	vsc_sata_interrupt(0, host_set, NULL);
+}
+
 
 static Scsi_Host_Template vsc_sata_sht = {
 	.module			= THIS_MODULE,
@@ -204,6 +232,8 @@
 	.dma_boundary		= ATA_DMA_BOUNDARY,
 	.slave_configure	= ata_scsi_slave_config,
 	.bios_param		= ata_std_bios_param,
+	.dump_quiesce		= vsc_dump_quiesce,
+	.dump_poll		= vsc_dump_poll,
 };
 
 

